File: src/create_measurements.py
#!/ usr / bin / env python3
"""
Generate measurements.txt for the 1 Billion Row Challenge.

Each line has the format: <station_name>;<temperature>
Temperature is a float with exactly one decimal place.
"""

import csv
import os
import random
import sys
import time
from pathlib import Path

#Default number of rows
DEFAULT_NUM_ROWS = 1_000_000_000

#Path configuration
SCRIPT_DIR = Path(__file__).parent
PROJECT_DIR = SCRIPT_DIR.parent
DATA_DIR = PROJECT_DIR / "data"
DATASETS_DIR = PROJECT_DIR / "datasets"
STATIONS_FILE = DATA_DIR / "weather_stations.csv"
OUTPUT_FILE = DATASETS_DIR / "measurements.txt"


def load_stations(filepath: Path) -> list[tuple[str, float]]:
    """Load weather stations from CSV file."""
    stations = []
    with open(filepath, "r", encoding="utf-8") as f:
        reader = csv.DictReader(f)
        for row in reader:
            name = row["station_name"]
            mean_temp = float(row["mean_temp"])
            stations.append((name, mean_temp))
    return stations


def generate_measurement(station: tuple[str, float]) -> str:
    """Generate a single measurement line."""
    name, mean_temp = station
#Generate temperature with normal distribution around mean
#Standard deviation of 10 degrees for realistic variation
    temp = random.gauss(mean_temp, 10.0)
#Clamp to valid range
    temp = max(-99.9, min(99.9, temp))
    return f"{name};{temp:.1f}\n"


def create_measurements(num_rows: int, output_path: Path) -> None:
    """Generate the measurements file."""
    print(f"Loading weather stations from {STATIONS_FILE}...")
    stations = load_stations(STATIONS_FILE)
    print(f"Loaded {len(stations)} weather stations")

    print(f"Generating {num_rows:,} measurements to {output_path}...")
    print("This may take a while for large datasets...")

#Ensure output directory exists
    output_path.parent.mkdir(parents=True, exist_ok=True)

    start_time = time.time()
    batch_size = 10_000_000  # Write in batches for better performance

    with open(output_path, "w", encoding="utf-8") as f:
        rows_written = 0
        while rows_written < num_rows:
#Generate a batch
            batch_count = min(batch_size, num_rows - rows_written)
            batch = [generate_measurement(random.choice(stations)) for _ in range(batch_count)]
            f.writelines(batch)
            rows_written += batch_count

#Progress update
            elapsed = time.time() - start_time
            progress = rows_written / num_rows * 100
            rate = rows_written / elapsed if elapsed > 0 else 0
            eta = (num_rows - rows_written) / rate if rate > 0 else 0
            print(
                f"\rProgress: {progress:.1f}% ({rows_written:,}/{num_rows:,}) "
                f"- {rate:,.0f} rows/sec - ETA: {eta:.0f}s",
                end="",
                flush=True,
            )

    elapsed = time.time() - start_time
    file_size = output_path.stat().st_size / (1024**3)  # Size in GB

    print(f"\n\nCompleted in {elapsed:.1f} seconds")
    print(f"File size: {file_size:.2f} GB")
    print(f"Output: {output_path}")


def main():
    num_rows = DEFAULT_NUM_ROWS

    if len(sys.argv) > 1:
        try:
            num_rows
= int(sys.argv[1]) except ValueError
    : print(f "Usage: {sys.argv[0]} [num_rows]")
          print(f "  num_rows: Number of measurements to generate (default: "
                  "{DEFAULT_NUM_ROWS:,})") sys.exit(1)

              create_measurements(num_rows, OUTPUT_FILE)

                  if __name__ == "__main__" : main()

File: src/create_measurements.cpp
#include <algorithm>
#include <atomic>
#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <random>
#include <sstream>
#include <string>
#include <thread>
#include <vector>

struct Station {
  std::string name;
  float mean_temp;
};

std::vector<Station> load_stations(const std::string &filepath) {
  std::vector<Station> stations;
  std::ifstream file(filepath);

  if (!file.is_open()) {
    std::cerr << "Error: Could not open " << filepath << std::endl;
    exit(1);
  }

  std::string line;
  std::getline(file, line); // Skip header

  while (std::getline(file, line)) {
    size_t comma = line.find(',');
    if (comma != std::string::npos) {
      Station s;
      s.name = line.substr(0, comma);
      s.mean_temp = std::stof(line.substr(comma + 1));
      stations.push_back(s);
    }
  }

  return stations;
}

void generate_chunk(const std::vector<Station> &stations, size_t num_rows,
                    std::string &output, unsigned int seed) {
  std::mt19937 rng(seed);
  std::uniform_int_distribution<size_t> station_dist(0, stations.size() - 1);

  output.reserve(num_rows * 20); // Estimate ~20 bytes per line

  char buf[64];

  for (size_t i = 0; i < num_rows; ++i) {
    const Station &s = stations[station_dist(rng)];

    // Normal distribution around mean temp with stddev of 10
    std::normal_distribution<float> temp_dist(s.mean_temp, 10.0f);
    float temp = temp_dist(rng);

    // Clamp to valid range
    temp = std::max(-99.9f, std::min(99.9f, temp));

    // Format with 1 decimal place
    int len = snprintf(buf, sizeof(buf), "%s;%.1f\n", s.name.c_str(), temp);
    output.append(buf, len);
  }
}

int main(int argc, char *argv[]) {
  size_t num_rows = 1'000'000'000; // Default 1 billion

  if (argc > 1) {
    num_rows = std::stoull(argv[1]);
  }

  const std::string stations_file = "data/weather_stations.csv";
  const std::string output_file = "datasets/measurements.txt";

  std::cout << "Loading weather stations..." << std::endl;
  std::vector<Station> stations = load_stations(stations_file);
  std::cout << "Loaded " << stations.size() << " stations" << std::endl;

  size_t n_threads = std::thread::hardware_concurrency();
  std::cout << "Using " << n_threads << " threads" << std::endl;
  std::cout << "Generating " << num_rows << " measurements..." << std::endl;

  auto start_time = std::chrono::high_resolution_clock::now();

  // Each thread generates its chunk into a string buffer
  std::vector<std::string> buffers(n_threads);
  std::vector<std::thread> threads;

  size_t rows_per_thread = num_rows / n_threads;
  size_t remainder = num_rows % n_threads;

  for (size_t i = 0; i < n_threads; ++i) {
    size_t thread_rows = rows_per_thread + (i < remainder ? 1 : 0);
    threads.emplace_back(generate_chunk, std::cref(stations), thread_rows,
                         std::ref(buffers[i]), (unsigned int)(i + 42));
  }

  // Wait for all threads
  for (auto &t : threads) {
    t.join();
  }

  auto gen_time = std::chrono::high_resolution_clock::now();
  auto gen_elapsed =
      std::chrono::duration<double>(gen_time - start_time).count();
  std::cout << "Generation completed in " << gen_elapsed << "s" << std::endl;

  // Write all buffers to file
  std::cout << "Writing to file..." << std::endl;
  FILE *f = fopen(output_file.c_str(), "w");
  if (!f) {
    std::cerr << "Error: Could not open output file" << std::endl;
    return 1;
  }

  for (size_t i = 0; i < n_threads; ++i) {
    fwrite(buffers[i].data(), 1, buffers[i].size(), f);
    buffers[i].clear();
    buffers[i].shrink_to_fit(); // Free memory as we go
  }
  fclose(f);

  auto end_time = std::chrono::high_resolution_clock::now();
  auto total_elapsed =
      std::chrono::duration<double>(end_time - start_time).count();

  std::cout << "Completed in " << total_elapsed << "s" << std::endl;
  std::cout << "Rate: " << (num_rows / total_elapsed / 1e6)
            << " million rows/sec" << std::endl;
  std::cout << "Output: " << output_file << std::endl;

  return 0;
}
